File: Python/Fraction to Recurring Decimal.py
class Solution:
    # @return a string
    def fractionToDecimal(self, numerator, denominator):
        if numerator == 0:
            return "0";
        ret = []
        if (numerator > 0) ^ (denominator > 0):
            ret.append("-")
        numerator = abs(numerator);
        denominator = abs(denominator)

        ret.append(str(numerator // denominator))
        if numerator % denominator == 0:
            return "".join(ret)
        else:
            ret.append(".")
        resultMap = {}

        remain = numerator % denominator
        while  remain != 0:
            if remain in resultMap:
                ret.insert(resultMap[remain] ,"(")
                ret.append(")")
                break
            resultMap[remain] = len(ret)
            remain *= 10
            ret.append(str(remain // denominator))
            remain %= denominator
        return "".join(ret)

File: Java/src/Fraction to Recurring Decimal.java
public class Solution {
    public String fractionToDecimal(int numerator, int denominator) {
        if (numerator == 0) return "0";
        if (denominator == 0) return "NaN";
        
        StringBuilder ret = new StringBuilder();
        ret.append(((numerator > 0) ^ (denominator > 0)) ? "-" : "");
        long numer = Math.abs((long)numerator);
        long denomin = Math.abs((long) denominator);
        
        ret.append(Long.toString(numerator/denominator));
        
        if (denomin % numer == 0) {
            return ret.toString();
        }
        else {
            ret.append(".");
        }
        
        HashMap<Long, Integer> map = new HashMap<>();
        Long remain = denomin % numer;
        while (remain != 0){
            if(map.containsKey(remain)){
                ret.insert(map.get(remain),"(");
                ret.append(")");
                break;
            }
            map.put(remain, ret.length());
            remain *= 10;
            ret.append(Long.toString(remain / denomin));
            remain %= denomin;
        }
        
        return ret.toString();
    }
}